File: 백준1012 유기농배추/복습/Main.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    static int M = 0; //배추밭 가로길이
    static int K = 0; //배추가 심어져 있는 위치의 개수
    private static int[][] map;
    private static boolean[][] isVisited;
    private static int N = 0; //배추밭 세로길이
    private static int result = 0; //지렁이개수
    private static int[] dx = {-1, 0, 1, 0};
    private static int[] dy = {0, -1, 0, 1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        for (int t = 0; t < T; t++) {
            result = 0;
            M = sc.nextInt();
            N = sc.nextInt();
            map = new int[N + 2][M + 2];
            isVisited = new boolean[N + 2][M + 2];
            K = sc.nextInt();
            for (int i = 0; i < K; i++) {
                int y = sc.nextInt() + 1;
                int x = sc.nextInt() + 1;
                map[x][y] = 1;
            }
            for (int i = 1; i <= N; i++) {
                for (int j = 1; j <= M; j++) {
                    if (!isVisited[i][j] && map[i][j] == 1) {
                        bfs(i, j);
                        result++;
                    }
                }
            }
            System.out.println(result);
        }
    }

    static void bfs(int x, int y) {
        Queue<Point> queue = new LinkedList<>();
        queue.offer(new Point(x, y));
        isVisited[x][y] = true;
        while (!queue.isEmpty()) {
            Point point = queue.poll();
            for (int i = 0; i < 4; i++) {
                int x2 = point.x + dx[i];
                int y2 = point.y + dy[i];
                if (map[x2][y2] == 1 && !isVisited[x2][y2]) {
                    isVisited[x2][y2] = true;
                    queue.offer(new Point(x2, y2));
                }
            }
        }
    }

    static class Point {
        int x;
        int y;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

}

File: 백준1012 유기농배추/복습/Main.kt
import java.util.*

internal var M = 0 //배추밭 가로길이
internal var K = 0 //배추가 심어져 있는 위치의 개수
private var map: Array<IntArray>? = null
private var isVisited: Array<BooleanArray>? = null
private var N = 0 //배추밭 세로길이
private var result = 0 //지렁이개수
private val dx = intArrayOf(-1, 0, 1, 0)
private val dy = intArrayOf(0, -1, 0, 1)


fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    val T = sc.nextInt()
    for (t in 0 until T) {
        result = 0
        M = sc.nextInt()
        N = sc.nextInt()
        map = Array(N + 2) { IntArray(M + 2) }
        isVisited = Array(N + 2) { BooleanArray(M + 2) }
        K = sc.nextInt()
        for (i in 0 until K) {
            val y = sc.nextInt() + 1
            val x = sc.nextInt() + 1
            map!![x][y] = 1
        }
        for (i in 1..N) {
            for (j in 1..M) {
                if (!isVisited!![i][j] && map!![i][j] == 1) {
                    bfs(i, j)
                    result++
                }
            }
        }
        println(result)
    }
}

fun bfs(x: Int, y: Int) {
    val queue = LinkedList<Point>()
    queue.offer(Point(x, y))
    isVisited!![x][y] = true
    while (!queue.isEmpty()) {
        val point = queue.poll()
        for (i in 0..3) {
            val x2 = point.x + dx[i]
            val y2 = point.y + dy[i]
            if (map!![x2][y2] == 1 && !isVisited!![x2][y2]) {
                isVisited!![x2][y2] = true
                queue.offer(Point(x2, y2))
            }
        }
    }
}

data class Point(var x: Int, var y: Int)


